arangod/Transaction/Methods.cpp

```diff
@@ -1148,151 +1273,169 @@ Future<OperationResult> transaction::Methods::insertLocal(
+      TRI_ASSERT(!showReplaced || oldRevisionId.isSet());
+      TRI_ASSERT(!showReplaced || previousDocumentBuilder->slice().isObject());
 
-      if (replicationType != ReplicationType::FOLLOWER) {
-        buildDocumentIdentity(collection.get(), resultBuilder, cid, keyString,
-                              docResult.revisionId(),
-                              prevDocResult.revisionId(),
-                              showReplaced ? &prevDocResult : nullptr,
-                              options.returnNew ? &docResult : nullptr);
-      }
+      key = newDocumentBuilder->slice().get(StaticStrings::KeyString);
+
+      buildDocumentIdentity(
+          *collection, resultBuilder, cid, key.stringView(), newRevisionId,
+          oldRevisionId, showReplaced ? previousDocumentBuilder.get() : nullptr,
+          options.returnNew ? newDocumentBuilder.get() : nullptr);
     }
+
+    if (!excludeFromReplication) {
+      TRI_ASSERT(newDocumentBuilder->slice().isObject());
+      // _id values are written to the database as VelocyPack Custom values.
+      // However, these cannot be transferred as Custom types, because the
+      // VelocyPack validator on the receiver side will complain about them.
+      // so we need to rewrite the document here to not include any Custom
+      // types.
+      // replicationData->add(newDocumentBuilder->slice());
+      basics::VelocyPackHelper::sanitizeNonClientTypes(
+          newDocumentBuilder->slice(), VPackSlice::noneSlice(),
+          *replicationData, transactionContextPtr()->getVPackOptions(), true,
+          true, false);
+    }
+
     return res;
   };
 
-  Result res;
   std::unordered_map<ErrorCode, size_t> errorCounter;
-  std::unordered_set<size_t> excludePositions;
 
+  replicationData->openArray(true);
   if (value.isArray()) {
-    VPackArrayBuilder b(&resultBuilder);
-    VPackArrayIterator it(value);
-    while (it.valid()) {
-      bool excludeFromReplication = false;
-      VPackSlice s = it.value();
-      TRI_IF_FAILURE("insertLocal::fakeResult1") {
-        res.reset(TRI_ERROR_DEBUG);
-        createBabiesError(replicationType == ReplicationType::FOLLOWER
-                              ? nullptr
-                              : &resultBuilder,
-                          errorCounter, res);
-        it.next();
-        continue;
+    resultBuilder.openArray();
+
+    for (VPackSlice s : VPackArrayIterator(value)) {
+      TRI_IF_FAILURE("insertLocal::fakeResult1") { res.reset(TRI_ERROR_DEBUG); }
+      if (res.ok()) {
+        res = workForOneDocument(s, true);
       }
-      res = workForOneDocument(s, true, excludeFromReplication);
       if (res.fail()) {
         createBabiesError(replicationType == ReplicationType::FOLLOWER
                               ? nullptr
                               : &resultBuilder,
                           errorCounter, res);
-      } else if (excludeFromReplication) {
-        excludePositions.insert(it.index());
+        res.reset();
```

Comment:
This `res.ok()` check seems superfluous since in the error case we always reset `res`.